### PR TITLE
HPCC-17068 Fix internal error when an action is associated with a persist name

### PR DESCRIPTION
**Pull request by ghalliday**
Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [x] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

## Testing:

Ran full compiler regression test.  For a few examples the persist hash changed - which means that persists may rebuild when migrating from a previous version of the platform.

### CODE REVIEW

**Comment by ghalliday** [comment]
Should this target 7.0 instead of 6.4 since it may cause some persists to rebuild?

**Comment by ghalliday** [comment]
@shamser please review

**Review by shamser — COMMENTED** [review]
@ghalliday Code looks good.  Just one comment regarding code documentation.

File: ecl/hql/hqlgram2.cpp [hunk 1]
```diff
@@ -2270,17 +2270,23 @@ void HqlGram::checkAllAssigned(IHqlExpression * originalRecord, IHqlExpression *
...
+    if (attr.queryExpr()->isConstant())
+    {
+        Owned<IHqlExpression> expr = attr.getExpr();
+        attr.setExpr(foldHqlExpression(expr), attr);
+    }
+    else
+        checkConstant(attr);
 }
 
 IHqlExpression * HqlGram::checkConstant(const attribute & errpos, IHqlExpression * expr)
```
Comment (ghalliday) [inline ecl/hql/hqlgram2.cpp]: @richardkchapman I think this provokes a general question we should discuss in the sprint/team meeting.  When should documentation be added to a function/class?

- When a new function or class is added?  I think we all agree yes.
- When a function is modified?  (The case for checkFoldConstant in this situation)
- When new code refers to an existing function?  (Vaguely the case in this situation for checkConstant - there are no new calls to the function.)

Arguments for always adding it are that it is the only way that the system is going to be documented.  If everyone knows that is what is expected then we will get into the habit of it, but it needs to be applied consistently.

Requiring any function that is changed (checkFoldConstant) to be documented would be reasonable - but do we want to make it a requirement?  Requiring documentation on anything that is called from modified code is likely to be too much.

There is also the additional problem if you modify code in a function you don't fully understand - but then maybe you shouldn't be modifying it until you fully understand it!

Comment (shamser) [inline ecl/hql/hqlgram2.cpp]: @ghalliday @richardkchapman if we add documentation to existing functions, the code documentation would not improve over time.

**Comment by richardkchapman** [comment]
Given the change to persists I think I would be inclined to target 7.0 - though if the internal error is causing significant issues and/or is not easy to work around I might be persuaded to change my mind

**Comment by richardkchapman** [comment]
@ghalliday See comment

**Comment by ghalliday** [comment]
@richardkchapman target changed to 7.0 - it is a fairly weird issue with a simple workaround.  Spotting what is wrong is slightly trickier, but I don't think it is very likely to occur.

**Pull request merged**